### ISSUE TITLE
Image size for training

### ISSUE THREAD

**Issue by wajahat57** [albanie/mcnFasterRCNN#5]
Hi
Its a very basic question.
Is it possible to train on variable image sizes?
Or, they have to be resized to 600 pxls smaller side?

In general, is it possible to train an isolated base network (such as VGG16 etc) on variable image size?

Best Regards
Wajahat

**Comment by albanie** [comment]
In general it should be fine.  However, the image size can't drop below a certain minimum, since each pooling operation in the trunk network reduces the spatial resolution of the features.  If the input image is too small, the tensors at `pool5` will become to small to perform roi pooling on top of them.

**Comment by wajahat57** [comment]
OK. But how about a bigger image during training. 
Don't the FC layers of the base network need a fixed size image during training to produce 1x1xN size tensors? If the image is bigger, than it will be more than SxTxN (S and T >1)?
How the loss will be computed?

**Comment by albanie** [comment]
No worries - the idea behind the roi pooling layer is that it converts arbitrary sized region proposals into fixed size feature vectors that are fed into the fully connected layers.  The caption for Figure 1 in the Fast R-CNN paper [here](https://arxiv.org/pdf/1504.08083.pdf) explains the general concept (Faster R-CNN uses the same mechanism).   This is one the primary differences between Faster R-CNN and detectors like SSD and YOLO, which don't use this a RoI pooling stage and therefore require a fixed input image size.

**Comment by wajahat57** [comment]
Great. Thanks again.
What is is this size:
frcn.divisions = [7, 7] ;
in:  faster_rcnn_init()
is it the ROI pooling size?

**Comment by albanie** [comment]
Yes, exactly.  The number of divisions defines the number of "bins" that the roi-pooling operation pools features into (this is what enables it to transform a variable size region proposal into a fixed size vector).  So for this model, every region proposal  be divided into a grid of 7 x 7 regions, each of which is max-pooled to give a single value.

**Comment by wajahat57** [comment]
Now I have got it. Fast-RCNN and even better the SPPnet paper explain this very well. Thanks for the pointers and explanation. 
Well, I need to connect a different base network to your Faster-RCNN architecture. I will need a little bit of help to understand the params. Faster-RCNN has to connect an ROIPooling removing the last MaxPooling (MP) layer of the base network and shape the output equal to the input feature map size expected by the first FC layer of the network. Right?

Now if there are Conv layers in between last MP layer the first FC layer, such as:
Conv4[3x3x64x128] > Relu4 >MP4 -->Conv5 [3x3x128x256] >Relu5 > FC6[5x14x256x512] >Relu6 >FC7[1x1x512x4096] >Relu7 >FC_Class[1x1x4096x1000]

So should I fix the params as follows:
      rpn.base = 'Relu4' ;
      frcn.base = 'Relu4' ; rpn.channels_in = 128 ;
      rpn.channels_out = 128 ;
      frcn.channels_in = 512 ;
      frcn.heads = {'FC6'} ;
      frcn.insert_dropout = false ;
      frcn.divisions = [5,14] ;
      frcn.tail = 'Relu7' ;
      freeze_decay = 1 ;
If so, in that case, will the Conv5>Relu5 block and FC7 be removed from the feature calculation?

**Comment by wajahat57** [comment]
Hi
When I use my own data for training, sometimes I get size mismatch error:
k =
    56
args =
  1×8 cell array
    [1×1×84×128 gpuArray]    [1×1×132×128 gpuArray]    'sigma'    [1]    'insideWeights'
    [1×1×132×128 double]    'outsideWeights'    [1×1×132×128 double]

as you can see, when using VOC dataset, all the gpuArrays are 1x1x84x128.
The image sizes are all 600Hx800W.
Can you tell where lies the mismatch?

**Comment by wajahat57** [comment]
Hi Albanie
Sorry to bother you again.
If I want to train on a different image size, such as 200x300, the even by changing the batchOpts.scale = 100 and therefore the resize factor, still the RPN cls loss and RPN bbox are 0 and NaN.
Can you suggest what else needs to be changed?
